  
# This master script will be built to replicate all plots from the Giant Virus paper which
# were generated by the R package Circlize


########################################################
########################################################

### Sections

# 1. Install and Load required Packages
# 2. Import and manipulate the data for extended figure 4
# 3. Replicate Extended figure 4
# TBD

########################################################
########################################################

# Section 1: Install and Load required Packages 

  install.packages("circlize")
  install.packages("GFF")
  install.packages("Rtools")
  install.packages("ape")
  
  # Commented these out because I'm not sure if they are really needed
  #install.packages("BiocManager")
  #BiocManager::install("genbankr")
  #library(BiocManager)
  
  library(circlize)
  library(ape)
  
########################################################
########################################################
  
# Section 2: Import and manipulate the data for extended figure 4 
  
  setwd("C:/Users/tyler/Desktop/test")
  
  
  # read idea from from https://www.rdocumentation.org/packages/ape/versions/5.4-1/topics/read.gff
  AUG_GEVE1 <-read.gff("AUG_eustigma.gff")
  AUG_GEVE2 <-read.gff("AUG_eustigma2.gff")
  PRG_GEVE1 <-read.gff("prodigal_eustigma")
  PRG_GEVE2 <-read.gff("prodigal_eustigma2")
  
  AUG_GEVE1 <- AUG_GEVE1[(AUG_GEVE1$type=="gene"),]
  PRG_GEVE1 <- PRG_GEVE1[(PRG_GEVE1$type=="gene"),]
  
  # get only the 3 coloumns seqid, start, and end. Idea from https://www.listendata.com/2015/06/r-keep-drop-columns-from-data-frame.html
  AUG_GEVE1 <- subset(AUG_GEVE1,select = +c(seqid,start,end))
  AUG_GEVE2 <- subset(AUG_GEVE2,select = +c(seqid,start,end))
  PRG_GEVE1 <- subset(PRG_GEVE1,select = +c(seqid,start,end))
  PRG_GEVE2 <- subset(PRG_GEVE2,select = +c(seqid,start,end))
  
 
 
  
  
########################################################
########################################################
  
  # Section 3: Replicate Extended figure 4 
  
  # code from https://jokergoo.github.io/circlize_book/book/graphics.html#fig:circlize-barplot chapter 3
  
  
  # Method 1 with a general plot Doesn't seem to be what I need
  circos.clear()
  circos.initialize(AUG_GEVE1$seqid,xlim = c(0, 10),sector.width=c(1514, ))
  circos.track(ylim = c(0, 1), panel.fun = function(x, y) {
    circos.barplot(value=1, bar_width=0.2,1, col = "brown")})
  title("C. eustigma GEVE 1")
  
  
  print(149280-145)
  print(unique(AUG_GEVE1$seqid))

  # Added sizes, still doesn't seem to be what I need
  circos.initialize(unique(AUG_GEVE1$seqid),xlim = c(0, 10),sector.width=c(10,20,30,40,50))
  circos.track(ylim = c(0, 1), panel.fun = function(x, y) {
    circos.barplot(value=1, bar_width=0.2,1, col = "brown")})
  title("C. eustigma GEVE 1")
  circos.clear()
  
  
  # Testing
  AUG_GEVE1 <- subset(AUG_GEVE1,select = +c(seqid,start,end,attributes,type))
  
  circos.par("start.degree" = 90)
  circos.genomicInitialize(AUG_GEVE1,plotTyper=NULL)
  circos.track(ylim = c(0, 1), bg.border = "black", track.height = 0.05)
  n = max(tapply(AUG_GEVE1$attributes,AUG_GEVE1$seqid, function(x) length(unique(x))))
  circos.genomicTrack(AUG_GEVE1, ylim = c(0.5, n + 0.5), 
                      panel.fun = function(region, value, ...) {
                        all_tx = unique(value$attributes)
                        for(i in seq_along(all_tx)) {
                          l = value$attributes == all_tx[i]
                          # for each transcript
                          current_tx_start = min(region[l, 1])
                          current_tx_end = max(region[l, 2])
                          circos.lines(c(current_tx_start, current_tx_end), 
                                       c(n - i + 1, n - i + 1), col = "#CCCCCC")
                          circos.genomicRect(region[1, , drop = FALSE], ytop = n - i + 1 + 0.4, 
                                             ybottom = n - i + 1 - 0.4, col = "orange", border = NA)
                        }
                      }, bg.border = NA, track.height = 0.4)
  circos.clear()
  
  
  
  # Method 2 with a genetic Initialize
  circos.par("start.degree" = 90)
  circos.genomicInitialize(AUG_GEVE1, plotType=NULL)
  circos.genomicTrackPlotRegion(AUG_GEVE1,ylim = c(0, 1),bg.border = "black", track.height = 0.1)
  circos.genomicTrackPlotRegion(PRG_GEVE1,ylim = c(0, 1),bg.border = "black", track.height = 0.1)
  
  title("C. eustigma GEVE 1")
  circos.clear()
  
 
  
  circos.initializeWithIdeogram(plotType = NULL)
  circos.track(ylim = c(0, 1), panel.fun = function(x, y) {
    chr = AUG_GEVE2.1$seqid
    xlim = AUG_GEVE2.1$start
    ylim = AUG_GEVE2.1$end
    circos.rect(xlim[1], 0, xlim[2], 1, col = rand_color(1))
    circos.text(mean(xlim), mean(ylim), chr, cex = 0.7, col = "white",
                facing = "inside", niceFacing = TRUE)
  }, track.height = 0.15, bg.border = NA)
  
  
  
  circos.initialize(unique(AUG_GEVE1.1$seqid),xlim = c(0, 10))
  circos.track(ylim = c(0, 1), panel.fun = function(x, y) {
    circos.barplot(value=1, bar_width=0.2,1, col = "brown",
                   sector.index = get.current.sector.index())
  })
  circos.track(ylim = c(0, 1), panel.fun = function(x, y) {
    circos.barplot(1, 1, col = "green")
  })
  circos.track(ylim = c(0, 1), panel.fun = function(x, y) {
    circos.barplot(1, 1, col = "purple")
  })
  title("C. eustigma GEVE 1")
  
  circos.initialize(letters[1:6], xlim = c(0, 10))
  par(xpd = NA)
  circos.track(ylim = c(0, 1), panel.fun = function(x, y) {
    circos.barplot(1, 1, col = "brown")
  })
  circos.track(ylim = c(0, 1), panel.fun = function(x, y) {
    circos.barplot(1, 1, col = "green")
  })
  circos.track(ylim = c(0, 1), panel.fun = function(x, y) {
    circos.barplot(1, 1, col = "purple")
  })
  title("C. eustigma GEVE 2")
  circos.clear()
  
########################################################
########################################################
  
  # Section 4: Testing things, no clue if its right  

  # code ideas from https://mran.microsoft.com/snapshot/2014-12-11/web/packages/circlize/vignettes/genomic_plot.pdf
  
  circos.clear()
  par(mar = c(1, 1, 1, 1))
  circos.par("start.degree" = 90)
  circos.initializeWithIdeogram(AUG_GEVE1.1,plotType=NULL)
  circos.genomicTrackPlotRegion(ylim = c(0, 1),bg.col=c("brown"),bg.border = NA, track.height = 0.05)
  circos.genomicTrackPlotRegion(PRG_GEVE1.1,ylim = c(0, 1),bg.col=c("green"),bg.border = NA, track.height = 0.05)
  
  
  circos.clear()
  par(mar = c(1, 1, 1, 1))
  circos.par("start.degree" = 90)
  circos.initializeWithIdeogram(AUG_GEVE2.1,plotType=NULL)
  circos.genomicTrackPlotRegion(ylim = c(0, 1),bg.col=c("brown"),bg.border = NA, track.height = 0.05)
  circos.genomicTrackPlotRegion(PRG_GEVE2.1,ylim = c(0, 1),bg.col=c("green"),bg.border = NA, track.height = 0.05)
  
  circos.clear()
  par(mar = c(1, 1, 1, 1))
  circos.par("start.degree" = 90)
  circos.initializeWithIdeogram(PRG_GEVE2.1,plotType=NULL)
  circos.genomicTrackPlotRegion(ylim = c(0, 1),bg.col=c("brown"),bg.border = NA, track.height = 0.05)
  circos.genomicTrackPlotRegion(AUG_GEVE2.1,ylim = c(0, 1),bg.col=c("green"),bg.border = NA, track.height = 0.05)
  
  #circos.clear()
  #par(mar = c(1, 1, 1, 1))
  #circos.par("start.degree" = 90)
  #circos.initializeWithIdeogram(plotType=NULL)
  #circos.genomicTrackPlotRegion(AUG_GEVE2.1,ylim = c(0, 1),bg.col=c("brown"),bg.border = NA, track.height = 0.05)
  #circos.genomicTrackPlotRegion(PRG_GEVE2.1,ylim = c(0, 1),bg.col=c("green"),bg.border = NA, track.height = 0.05)
 
  n = max(sapply(test2, length))
  circos.genomicTrackPlotRegion(ylim = c(0.5, n + 0.5),
                                panel.fun = function(region, value, ...) {
                                  gn = get.cell.meta.data("sector.index")
                                  tr = test2[[gn]] # all transcripts for this gene
                                  for(i in seq_along(tr)) {
                                    # for each transcript
                                    current_tr_start = min(tr[[i]]$start)
                                    current_tr_end = max(tr[[i]]$end)
                                    circos.lines(c(current_tr_start, current_tr_end),
                                                 c(n - i, n - i), col = "#CCCCCC")
                                    circos.genomicRect(tr[[i]], ytop = n - i + 0.4,
                                                       ybottom = n - i - 0.4, col = "orange", border = NA)
                                  }
                                }, bg.border = NA, track.height = 0.3)
  circos.clear()