# This master script will be built to replicate all plots from the Giant Virus paper which
# were generated by the R package Circlize

########################################################
########################################################

##### Sections

# 1. Install and Load required Packages
# 2. Import and manipulate the data for extended figure 4
# 3. Replicate Extended figure 4
# TBD

########################################################
########################################################

# Section 1: Install and Load required Packages 

  install.packages("circlize")
  install.packages("GFF")
  install.packages("Rtools")
  install.packages("ape")
  
  library(circlize)
  library(ape)
  library(dplyr)
  
########################################################
########################################################
  
# Section 2: Import and manipulate the data for extended figure 4 
  
  setwd("C:/Users/tyler/Desktop/test1")
  
  # read idea from from https://www.rdocumentation.org/packages/ape/versions/5.4-1/topics/read.gff
  AUG_GEVE1 <-read.gff("AUG_eustigma.gff")
  AUG_GEVE2 <-read.gff("AUG_eustigma2.gff")
  PRG_GEVE1 <-read.gff("prodigal_eustigma")
  PRG_GEVE2 <-read.gff("prodigal_eustigma2")
  
  #AUG_GEVE1 <- AUG_GEVE1[(AUG_GEVE1$type=="gene"),]
  #PRG_GEVE1 <- PRG_GEVE1[(PRG_GEVE1$type=="gene"),]
  #AUG_GEVE2 <- AUG_GEVE2[(AUG_GEVE2$type=="gene"),]
  #PRG_GEVE2 <- PRG_GEVE2[(PRG_GEVE2$type=="gene"),]
  
  # get only the 4 coloumns seqid, start, end and attributes. Idea from https://www.listendata.com/2015/06/r-keep-drop-columns-from-data-frame.html
  AUG_GEVE1 <- subset(AUG_GEVE1,select = +c(seqid,start,end,attributes))
  AUG_GEVE2 <- subset(AUG_GEVE2,select = +c(seqid,start,end,attributes))
  PRG_GEVE1 <- subset(PRG_GEVE1,select = +c(seqid,start,end,attributes))
  PRG_GEVE2 <- subset(PRG_GEVE2,select = +c(seqid,start,end,attributes))
  
  # Convert factors to character
  AUG_GEVE1$seqid <- as.character(AUG_GEVE1$seqid)
  PRG_GEVE1$seqid <- as.character(PRG_GEVE1$seqid)
  AUG_GEVE2$seqid <- as.character(AUG_GEVE2$seqid)
  PRG_GEVE2$seqid <- as.character(PRG_GEVE2$seqid)
  
  
  
  # Remove the prodigal genes which overlap with Augustus for GEVE1
  # For whatever stupid reason I have to run this multiple times and even then it doesn't work right
  n=1  
  for(i in PRG_GEVE1$seqid){
    m=1
    if(n<nrow(PRG_GEVE1)){
      for(j in AUG_GEVE1$seqid){
        if(m<nrow(AUG_GEVE1)){
          if(i == j){
            # Remove a Prodigal predicted gene if it lies within an Augustus Predicted gene
            if((PRG_GEVE1[n,2]>=AUG_GEVE1[m,2] && PRG_GEVE1[n,3]<=AUG_GEVE1[m,3])||
               (PRG_GEVE1[n,2]<=AUG_GEVE1[m,2] && PRG_GEVE1[n,3]>=AUG_GEVE1[m,3])){
              PRG_GEVE1 <- PRG_GEVE1[-c(n),]
              rownames(PRG_GEVE1) <- NULL
            }
          }
        }
        m=m+1
      }
    }
    n=n+1
  }
  
########################################################
########################################################
  
  # Section 3: Replicate Extended figure 4 
  
  # idea from section 9.3 of https://jokergoo.github.io/circlize_book/book/initialize-genomic-plot.html#initialize-with-general-genomic-category
  
  # Create GEVE1 plot
  circos.par("start.degree" = 90)
  circos.genomicInitialize(AUG_GEVE1,plotType=NULL)
  a1 = max(tapply(AUG_GEVE1$attributes,AUG_GEVE1$seqid, function(x) length(unique(x))))
  circos.genomicTrack(AUG_GEVE1, ylim = c(0.5, a1 + 0.5), 
                      panel.fun = function(region, value, ...) {
                        all_tx = unique(value$attributes)
                        for(i in seq_along(all_tx)) {
                          l = value$attributes == all_tx[i]
                          current_tx_start = min(region[l, 1])
                          current_tx_end = max(region[l, 2])
                          circos.lines(x=c(current_tx_start, current_tx_end),y=c(a1,a1),col = "brown", area=TRUE)
                        }
                      }, bg.border = "black", track.height = 0.1)
  p1 = max(tapply(PRG_GEVE1$attributes,PRG_GEVE1$seqid, function(x) length(unique(x))))
  circos.genomicTrack(PRG_GEVE1, ylim = c(0.5, p1 + 0.5), 
                      panel.fun = function(region, value, ...) {
                        all_tx2 = unique(value$attributes)
                        for(i in seq_along(all_tx2)) {
                          l = value$attributes == all_tx2[i]
                          # for each transcript
                          current_tx_start = min(region[l, 1])
                          current_tx_end = max(region[l, 2])
                          circos.lines(x=c(current_tx_start, current_tx_end),y=c(p1,p1), col = "green", area=TRUE)
                          #circos.link(sector.index1 = get.current.sector.index(),c(0,1),sector.index2 = "BEGY01000127.1",c(0,1))
                        }
                      }, bg.border = "black", track.height = 0.1)
  circos.genomicTrack(ylim = c(0.5, p1 + 0.5), bg.border = "black", track.height = 0.1)
  title("C. eustigma GEVE 1")
  circos.clear()
