  
# This master script will be built to replicate all plots from the Giant Virus paper which
# were generated by the R package Circlize


########################################################
########################################################

### Sections

# 1. Install and Load required Packages
# 2. Import and manipulate the data for extended figure 4
# 3. Create the circular plots for C. eustigma's GEVEs
# TBD

########################################################
########################################################

# Section 1: Install and Load required Packages 

  install.packages("circlize")
  install.packages("GFF")
  
  library(circlize)
  library(dplyr)
  library(ape)
  library(readxl)
  
########################################################
########################################################
  
# Section 2: Import and manipulate the data for extended figure 4 
  
  setwd("C:/Users/tyler/Desktop/test data")
  
  # read the sequence size info
  seq_info <-read_excel("Sequence Sector Info.xlsx")
  seq_list<-split(seq_info, seq_info$GEVE)
  Ceu_geve1_sectors<- seq_list[[1]]
  Ceu_geve1_sectors<- Ceu_geve1_sectors[,c(1:3)]
  Ceu_geve2_sectors<- seq_list[[2]]
  Ceu_geve2_sectors<- Ceu_geve2_sectors[,c(1:3)]
  
  
  # read idea from from https://www.rdocumentation.org/packages/ape/versions/5.4-1/topics/read.gff
  # read the output files from augustus and prodigal predictions into a dataframe
  AUG_GEVE1 <-read.gff("AUG_eustigma.gff")
  PRG_GEVE1 <-read.gff("prodigal_eustigma")
  AUG_GEVE2 <-read.gff("AUG_eustigma2.gff")
  PRG_GEVE2 <-read.gff("prodigal_eustigma2")
  NCVOG_Hits <- read_excel("NCVOG.xlsx")
  #NCVOG_Hits<-read.csv(file="NCVOG Hits.csv",sep=",", header = T)
  
  # Convert factors to character
  AUG_GEVE1$seqid <- as.character(AUG_GEVE1$seqid)
  PRG_GEVE1$seqid <- as.character(PRG_GEVE1$seqid)
  
  # remove hits which are not genes
  #AUG_GEVE1 <- AUG_GEVE1[(AUG_GEVE1$type=="gene"),]
  
  # get only the 4 coloumns seqid, start, end and attributes. Idea from https://www.listendata.com/2015/06/r-keep-drop-columns-from-data-frame.html
  AUG_GEVE1 <- subset(AUG_GEVE1,select = +c(seqid,start,end))
  PRG_GEVE1 <- subset(PRG_GEVE1,select = +c(seqid,start,end))
  AUG_GEVE2 <- subset(AUG_GEVE2,select = +c(seqid,start,end))
  PRG_GEVE2 <- subset(PRG_GEVE2,select = +c(seqid,start,end))
  NCVOG_Hits <-subset(NCVOG_Hits,select = +c(seqid,start,end))

  # Remove the prodigal genes which overlap with Augustus for GEVE1
  # For whatever reason I have to run this 3 times
  
  for (row in 1:nrow(PRG_GEVE1)) {
    P_start = PRG_GEVE1[row,2]
    P_end = PRG_GEVE1[row,3]
    for (checks in 1:nrow(AUG_GEVE1)){
      A_start=AUG_GEVE1[checks,2]
      A_end = AUG_GEVE1[checks,3]
      if (PRG_GEVE1[row,1]==AUG_GEVE1[checks,1]){
        if((P_start>=A_start && P_end<=A_end) || (P_start<=A_start && P_end>=A_end)){
          PRG_GEVE1 <- PRG_GEVE1[-c(row),]
        }
      }
    }
    rownames(PRG_GEVE1) <- NULL
  }
  
  for (row in 1:nrow(PRG_GEVE1)) {
    P_start = PRG_GEVE1[row,2]
    P_end = PRG_GEVE1[row,3]
    for (checks in 1:nrow(NCVOG_Hits)){
      A_start=NCVOG_Hits[checks,2]
      A_end = NCVOG_Hits[checks,3]
      if (PRG_GEVE1[row,1]==AUG_GEVE1[checks,1]){
        if((P_start<=A_start && P_end>=A_end) || (P_start>=A_start && P_end<=A_end)){
          NCVOG_Hits <- NCVOG_Hits[-c(checks),]
        }
        if(P_start<=A_start && P_end>=A_start){
          NCVOG_Hits <- NCVOG_Hits[-c(checks),]
        }
        if(P_start>=A_start && P_start<=A_end){
          NCVOG_Hits <- NCVOG_Hits[-c(checks),]
        }
      }
    }
    rownames(NCVOG_Hits) <- NULL
  }
  
  png("Plot3.png", width = 4, height = 4, units = 'in', res = 800)
  
  
########################################################
########################################################
  
  # Section 3: Replicate Extended figure 4 
  
  # idea from section 9.3 of https://jokergoo.github.io/circlize_book/book/initialize-genomic-plot.html#initialize-with-general-genomic-category
  
  # Create GEVE1 plot
  circos.clear()
  circos.par(start.degree=90, track.height=0.1)
  
  circos.genomicInitialize(Ceu_geve1_sectors, plotType = NULL)
  circos.genomicTrack(AUG_GEVE1, ylim=c(0,1), track.height=0.1, bg.lwd=0.4, bg.border="gray", panel.fun=function(region,value,...){ circos.genomicRect(region,value, col=("saddlebrown"),border=NA,...)})
  circos.genomicTrack(PRG_GEVE1, ylim=c(0,1), track.height=0.1, bg.lwd=0.4, bg.border="gray", panel.fun=function(region,value,...){ circos.genomicRect(region,value, col=("darkgreen"),border=NA,...)})
  circos.genomicTrack(NCVOG_Hits, ylim=c(0,1), track.height=0.1, bg.lwd=0.4, bg.border="gray", panel.fun=function(region,value,...){ circos.genomicRect(region,value, col=("navy"),border="navy",...)})
  title("C. eustigma GEVE 1")
  
  dev.off()

  
  # Create GEVE2 plot
  circos.clear()
  circos.par(start.degree=90, track.height=0.1)
  
  circos.genomicInitialize(Ceu_geve2_sectors, plotType = NULL)
  circos.genomicTrack(AUG_GEVE2, ylim=c(0.5,2.5), track.height=0.1, bg.lwd=0.4, bg.border="gray47", panel.fun=function(region,value,...){ circos.genomicRect(region,value, col=("brown"),border=NA,...)})
  circos.genomicTrack(PRG_GEVE2, ylim=c(0.5,2.5), track.height=0.1, bg.lwd=0.4, bg.border="gray47", panel.fun=function(region,value,...){ circos.genomicRect(region,value, col=("darkgreen"),border=NA,...)})
  circos.genomicTrack(NCVOG_GEVE2, ylim=c(0.5,2.5), track.height=0.1, bg.lwd=0.4, bg.border="gray47", panel.fun=function(region,value,...){ circos.genomicRect(region,value, col=("purple"),border=NA,...)})
  title("C. eustigma GEVE 2")
  